Allow for restart without using parent/child processes.

jaharkes opened the pull request:
**Description:**
Simplify restarts using just the 'exec' part of fork and exec.

Assuming that we normally correctly shut down running threads and release resources, we just do some minimal scrubbing of open file descriptors and child processes which would normally stay around across an exec() boundary.

In case exec fails we return 100, which means that systemd could be set up to handle the restart automatically if we set `RestartForceExitStatus=100` in the systemd.service file.

Only tested on Linux, should work on MacOS X and aside for not cleaning up stray child processes "should" work on Windows.

**Related issue (if applicable):** #

**Example entry for `configuration.yaml` (if applicable):**

``` yaml

```

**Checklist:**

If the code does not interact with devices:
- [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
- [ ] Tests have been added to verify that the new code works.

jaharkes committed:
Allow for restart without using parent/child processes.

Assuming that we normally correctly shut down running threads and
release resources, we just do some minimal scrubbing of open file
descriptors and child processes which would stay around across an
exec() boundary.

balloob commented on homeassistant/__main__.py:
```diff
@@ -256,42 +250,61 @@ def open_browser(event):
+    # systemd will restart us when RestartForceExitStatus=100 is set in the
+    # systemd.service file.
+    sys.stderr.write("Restarting Home-Assistant\n")
+    exe = get_executable()
```
Alternative, we could get this using `sys.executable`

jaharkes replied:
I figured if multiprocessing is part of the standard library and is not using that function, it must have a reason for it.

balloob commented on homeassistant/__main__.py:
```diff
@@ -256,42 +250,61 @@ def open_browser(event):
 
         hass.bus.listen_once(EVENT_HOMEASSISTANT_START, open_browser)
 
+    print('Starting Home-Assistant')
```
Is this print intentional ?

jaharkes replied:
This is after we have loaded the configuration and are about to start the loop where we set up handlers and respond to SIGINT/SIGHUP/SIGTERM signals. It was helpful for me to know but I can remove it.

balloob commented:
It didn't work on Mac. Looks like we killed ourself?

![screen shot 2016-04-10 at 21 04 48](https://cloud.githubusercontent.com/assets/1444314/14416524/ed011204-ff5f-11e5-9241-4724b9b91c71.png)

jaharkes commented:
Yes indeed. Well technically we shouldn't really have any child processes if we cleaned up things right, and we can't kill them on windows anyway so maybe we could just ignore that bit. Worst case some zombies will hang around until the process is stopped for real and init can collect them.

jaharkes committed:
Use sys.executable instead of multiprocessing.spawn.get_executable()

balloob commented on homeassistant/__main__.py:
```diff
@@ -259,39 +251,33 @@ def open_browser(event):
-        except KeyboardInterrupt:
-            return False
+    # Try to not leave behind open filedescriptors with the emphasis on try.
+    os.closerange(3, 1024)
```
This line is causing OS X to kill the process. Commenting it out and I can safely restart.

jaharkes replied:
What? cleaning up file descriptors kills the process? I would not have expected that. I'll look into that a bit more to see why that is happening and maybe make a special case for macos. Generally closing fd is a good thing because it also gets rid of open sockets and such.

I can probably revert 705255d then and at least signal (child) processes that are still hanging out in our process group.

jaharkes replied:
From the [chromium bug tracker](https://chromium.googlesource.com/chromium/src/+/6111004802718e6566262f82699d0ab0ed4e53f3)
```
OSX 10.9 introduced a kernel facility which allows tagging file
descriptors to prevent various operations unless the caller knows the
key.  libdispatch uses this facility to tag a file descriptor against
close() or dup(), and this test does both.  Unfortunately, the failure
case is to raise a kernel exception and hang the child process rather
than return an error code.
```

jaharkes committed:
Limit how many file descriptors we try to close.

Don't even try to close on OSX/Darwin until we figure out how to
recognize guarded fds because the kernel will yell at us, and kill
the process.

jaharkes committed:
Use the close on exec flag on MacOS to clean up.

balloob commented:
Can confirm that it works on OS X. Let's get a Windows user to test this out.

Bart274 commented:
I just tested it, both the stop and the restart service aren't working anymore, while they still do in the dev version

balloob commented:
What's happening instead, console hangs?

On Tue, Apr 12, 2016, 07:33 Bart274 <notifications@github.com> wrote:

> I just tested it, both the stop and the restart service aren't working
> anymore, while they still do in the dev version
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly or view it on GitHub
> <https://github.com/balloob/home-assistant/pull/1793#issuecomment-208936059>
>

jaharkes commented:
It just kept running normally which makes me think it got stuck waiting for threads to stop. There are no significant changes in homeassistant.core that I can see breaking stop as well.

jaharkes commented:
now it did run a bunch of pip installs because of the lib->deps directory change, so that could have affected windows' ability to exit the main loop.

Bart274 commented:
That's the thing, nothing happens. It gets logged that the restart service is being called but then HA keeps working like it did before.

> Op 12 apr. 2016 om 16:48 heeft Paulus Schoutsen <notifications@github.com> het volgende geschreven:
> 
> What's happening instead, console hangs?
> 
> On Tue, Apr 12, 2016, 07:33 Bart274 <notifications@github.com> wrote:
> 
> > I just tested it, both the stop and the restart service aren't working
> > anymore, while they still do in the dev version
> >
> > —
> > You are receiving this because you commented.
> >
> >
> > Reply to this email directly or view it on GitHub
> > <https://github.com/balloob/home-assistant/pull/1793#issuecomment-208936059>
> >
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
>

Bart274 commented:
I'll try it again tomorrow and I'll report back 😉 

> Op 12 apr. 2016 om 16:51 heeft Jan Harkes <notifications@github.com> het volgende geschreven:
> 
> now it did run a bunch of pip installs because of the lib->deps directory change, so that could have affected windows' ability to exit the main loop.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
>

balloob commented:
Found this bug report when Googling around about this PR: http://bugs.python.org/issue9148

balloob commented:
It doesn't seem like there is a proper fix under Windows 😞 

We could limit this PR to only act like this on OS X + Linux and not support restarting on Windows. The forking is causing issues with double running instances (breaking HTTP). We could I guess also be more strict about ctrl+C and always try to do a clean exit.

jaharkes commented:
It also makes it harder to debug with pdb etc. If the exec doesn't work, and we can explicitly disable it on windows, we return with exit code 100, so a nanny type process could handle restarting in that case. I was thinking of adding a startup flag (--runner ?) to enable this because this functionality is provided by systemd/supervisord and various other daemon runners.
It might be possible to create a really trivial 10 line daemon-runner that windows users can use if they want restart functionality and only expose exit functionality otherwise. I'll try to whip something up tonight.

jaharkes committed:
Introduce a small process runner to handle restart on windows.

balloob commented:
This looks good (as far as that I get it!). We should again make sure this runs on Windows.

Do you think we should try to get this into 0.18 or merge it right after the release so it can be vetted?

Christoph-Wagner commented:
I'm getting this on windows:

```
Traceback (most recent call last):
INFO:homeassistant.components.http:Starting web interface at http://0.0.0.0:80
  File "C:\Program Files (x86)\Python35-32\Scripts\hass-script.py", line 9, in <module>
INFO:homeassistant.core:Timer:starting
    load_entry_point('homeassistant==0.18.0.dev0', 'console_scripts', 'hass')()
  File "g:\malazan\homeassistant\home-assistant-code\homeassistant\__main__.py", line 392, in main
    exit_code = setup_and_run_hass(config_dir, args)
  File "g:\malazan\homeassistant\home-assistant-code\homeassistant\__main__.py", line 295, in setup_and_run_hass
    exit_code = int(hass.block_till_stopped())
  File "g:\malazan\homeassistant\home-assistant-code\homeassistant\core.py", line 97, in block_till_stopped
    signal.signal(signal.SIGHUP, restart_homeassistant)
AttributeError: module 'signal' has no attribute 'SIGHUP'
```

jaharkes commented:
Clearly not 0.18 material. I guess at some point I am going to have to start looking to install windows in a VM or something so I can iterate more quickly.

balloob commented:
Microsoft has your back with ready to download images:
https://developer.microsoft.com/en-us/microsoft-edge/tools/vms/

On Wed, Apr 20, 2016 at 12:26 PM, Jan Harkes <notifications@github.com>
wrote:

> Clearly not 0.18 material. I guess at some point I am going to have to
> start looking to install windows in a VM or something so I can iterate more
> quickly.
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/home-assistant/home-assistant/pull/1793#issuecomment-212568800>
>



-- 
PaulusSchoutsen.nl
It's nice to be important but it's more important to be nice.

jaharkes committed:
Handle missing signal.SIGHUP on Windows.

nkgilley commented:
My Docker image won't start with this:

```log
root@Tower:~# docker run -it homeassistant/home-assistant:dev
Traceback (most recent call last):
  File "/usr/local/lib/python3.4/runpy.py", line 170, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/local/lib/python3.4/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/usr/src/app/homeassistant/__main__.py", line 406, in <module>
    sys.exit(main())
  File "/usr/src/app/homeassistant/__main__.py", line 396, in main
    os.setpgid(0, 0)
PermissionError: [Errno 1] Operation not permitted
```

jaharkes commented:
Interesting, it must already be a process leader when launched in docker. We can catch and ignore that permission error.

jaharkes commented:
This should ignore the permission error, but I would like to know if restart still works as expected in a docker container after this patch is applied.

https://github.com/home-assistant/home-assistant/compare/ignore-setpgid-error

robbiet480 commented:
Getting the same issue as @nkgilley. I am not in Docker but a virtualenv.

robbiet480 commented:
This issue has been fixed by @jaharkes's #2118 PR.
